### PR TITLE
feat(FEC-7804): implement tooltips

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

Implemented tooltip support - 

- contains 8 types of tooltips (default top)
- boundary check and fallback to other types in case of invalid type
- multiline and max-width support (default 240px. none will make it one line)
- next and prev tooltips will not be shown in case the poster preview is shown
- added debounce util to limit a burst of function calls

Solves (FEC-7804)

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/components/cast/cast.js [hunk 1]
```diff
@@ -58,17 +61,21 @@ class Cast extends Component {
...
-          onKeyDown={e => {
-            if (e.keyCode === KeyMap.ENTER) {
-              this.props.updateBackdropVisibility(true);
-              this.props.player.startCasting().catch(() => this.props.updateBackdropVisibility(false));
-            }
-          }}>
-          <google-cast-launcher className={style.castButton} tabIndex="0" />
-        </div>
+        <Tooltip label={this.props.castText}>
+          <div
```
Comment (Yuvalke) [inline src/components/cast/cast.js]: why doesn't it button?

File: src/components/play-pause/play-pause.js [hunk 2]
```diff
@@ -57,14 +63,11 @@ class PlayPause extends Component {
   render(props: any): React$Element<any> | void {
     const controlButtonClass = this.props.isPlayingAdOrPlayback ? [style.controlButton, style.isPlaying].join(' ') : style.controlButton;
     const isStartOver = props.isPlaybackEnded && !this.props.adBreak;
+    const localization = isStartOver ? this.props.startOverText : this.props.isPlayingAdOrPlayback ? this.props.pauseText : this.props.playText;
```
Comment (Yuvalke) [inline src/components/play-pause/play-pause.js]: Split it to two rows, we don't use inline if in inline if. and more valuable name.
```
const buttonStateText = this.props.isPlayingAdOrPlayback ? this.props.pauseText : this.props.playText;
const buttonLabelText = isStartOver ? this.props.startOverText : buttonText;
```

File: src/components/playlist-button/playlist-button.js [hunk 3]
```diff
@@ -46,44 +53,57 @@ class PlaylistButton extends Component {
...
-                <Icon type={IconType.Prev} />
-              </div>
-            ) : (
-              <div>
-                <Icon type={IconType.Next} />
-              </div>
-            )}
-          </button>
-        </Localizer>
+        {showPreview ? (
```
Comment (Yuvalke) [inline src/components/playlist-button/playlist-button.js]: don't you need tooltip here as well?

File: src/components/settings/settings.js [hunk 4]
```diff
@@ -211,13 +212,15 @@ class Settings extends Component {
...
     return (
       <div ref={c => (this._controlSettingsElement = c)} className={[style.controlButtonContainer, style.controlSettings].join(' ')}>
-        <button
-          tabIndex="0"
-          aria-label={props.buttonAriaLabel}
-          className={this.state.smartContainerOpen ? [style.controlButton, style.active].join(' ') : style.controlButton}
-          onClick={() => this.onControlButtonClick()}>
-          <Icon type={IconType.Settings} />
-        </button>
+        <Tooltip label={this.props.buttonLabel}>
```
Comment (Yuvalke) [inline src/components/settings/settings.js]: make compatible with line 218

Comment (Yuvalke) [reply src/components/settings/settings.js]: I mean with `this` or without

File: src/components/tooltip/tooltip.js [hunk 5]
```diff
@@ -10,6 +35,159 @@ import {h, Component} from 'preact';
...
+    this.setState({showTooltip: false, marginTop: 0, marginLeft: 0});
+  }
+
+  /**
+   * gets the tooltip position based on component prop
+   *
+   * @returns {void}
+   * @memberof Tooltip
+   */
+  setMargins() {
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: why don't you create css class and just set the correct class? I think it's too much style work on js.

File: src/components/tooltip/tooltip.js [hunk 6]
```diff
@@ -10,6 +35,159 @@ import {h, Component} from 'preact';
...
+
+  /**
+   * checks if the current tooltip type is within the player boundaries
+   * @memberof Tooltip
+   * @returns {string} is in boundaries
+   */
+  isToolTipInBoundaries(): boolean {
+    const tooltipBoundingRect = this.textElement.getBoundingClientRect();
+    const playerContainerRect = this.props.playerClientRect;
+    const MARGIN = 10;
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: You have here constant like margin, that you are using it on css as well.
Could it be in css?

File: src/components/tooltip/_tooltip.scss [hunk 7]
```diff
@@ -1,31 +1,122 @@
...
+    &:after {
+      content: '';
+      position: absolute;
+      border-width: 5px;
+      border-style: solid;
+      border-radius: 3px;
+      color: $tooltip-bg-color;
+      z-index: -1;
+    }
+    &.tooltip-top {
```
Comment (Yuvalke) [inline src/components/tooltip/_tooltip.scss]: the constant here should be same for left and right, bottom and top right?
Could it be in constant ?

File: src/components/tooltip/tooltip.js (hunk 6 above)
Comment (OrenMe) [reply src/components/tooltip/tooltip.js]: magic numbers should be in the top of the file as CAPITAL snake case

File: src/components/cast/cast.js (hunk 1 above)
Comment (RoyBregman) [reply src/components/cast/cast.js]: This is old code - when working on the accessibility i tried changing it to button but the css was ruined so i rolled it back

File: src/components/play-pause/play-pause.js (hunk 2 above)
Comment (RoyBregman) [reply src/components/play-pause/play-pause.js]: i agree that inline if inside another inline if is not readable - but anyway the result is just one const not two. i will change it and see if you agree also

File: src/components/playlist-button/playlist-button.js (hunk 3 above)
Comment (RoyBregman) [reply src/components/playlist-button/playlist-button.js]: No. when there is a preview it is ugly and unnecessary to have also a tooltip

File: src/components/settings/settings.js (hunk 4 above)
Comment (RoyBregman) [reply src/components/settings/settings.js]: fixed

File: src/components/tooltip/_tooltip.scss (hunk 7 above)
Comment (RoyBregman) [reply src/components/tooltip/_tooltip.scss]: agree, fixed

File: src/components/tooltip/tooltip.js (hunk 5 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: Yep :) done

File: src/components/tooltip/tooltip.js (hunk 6 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: This is not the same margin as in the css - this represents the player margins (So the tooltip will reach until the last pixels). I renamed it to PLAYER_MARGIN and moved to the top like Oren mentioned.

File: src/components/tooltip/tooltip.js [hunk 8]
```diff
@@ -10,6 +36,116 @@ import {h, Component} from 'preact';
...
+
+  /**
+   * default component props
+   * @type {Object}
+   * @memberof Tooltip
+   */
+  static defaultProps = {
+    type: ToolTipType.Top,
+    isMultiline: true,
+    maxWidth: '240px'
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: create const

File: src/components/tooltip/tooltip.js [hunk 9]
```diff
@@ -18,15 +154,23 @@ class Tooltip extends Component {
...
-    let className = [style.tooltip];
-    if (props.out) className.push(style.out);
+    const className = [style.tooltipLabel, style[`tooltip-${this.state.type}`]];
+    this.state.showTooltip ? className.push(style.show) : className.push(style.hide);
 
     return (
-      <div className={className.join(' ')} style={props.left ? {left: props.left} : ''}>
+      <div className={style.tooltip} onMouseOver={() => this.onMouseOver()} onMouseLeave={() => this.onMouseLeave()}>
+        <span
+          style={{maxWidth: props.isMultiline ? props.maxWidth : 'none'}}
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: when this state change?

File: src/components/tooltip/tooltip.js [hunk 10]
```diff
@@ -10,6 +36,116 @@ import {h, Component} from 'preact';
...
+  }
+
+  /**
+   * checks if after the render the tooltip is within boundaries of the player
+   * if not it will try to set a new type which will be checked after the next render
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentDidUpdate(): void {
+    if (this.state.showTooltip && !this.isToolTipInBoundaries()) {
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: Don't you can test the all possible tooltips and select the one which can be used by order?
Here you make array of non valid type but the playerClientRect can change and they will be relevant again

File: src/components/tooltip/_tooltip.scss [hunk 11]
```diff
@@ -1,31 +1,129 @@
...
+      bottom: $tooltip-margin;
+      left: 50%;
+      transform: translate(-50%);
+      &:after {
+        bottom: -$tooltip-arrow-height;
+        left: 50%;
+        transform: translate(-50%) rotate(45deg) skew(5deg, 5deg);
+      }
+    }
+    &.tooltip-top-left {
```
Comment (yairans) [inline src/components/tooltip/_tooltip.scss]: in generally `.tooltip-top-left` should be a combination of `.tooltip-top` and `.tooltip-left` as we do in the watermark component. does it applicable here?
any way, `.tooltip-top-left` and `.tooltip-top-right` should use `.tooltip-top` rules. only overide the different rule. 
try to reuse the rules as much as possible

File: src/components/tooltip/tooltip.js [hunk 12]
```diff
@@ -1,6 +1,32 @@
...
+/**
+ * mapping state to props
+ * @param {*} state - redux store state
+ * @returns {Object} - mapped state to this component
+ */
+const mapStateToProps = state => ({
+  playerClientRect: state.shell.playerClientRect
+});
+
+export const TOOLTIP_SHOW_TIMEOUT: number = 250;
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: there isn't. i will remove the export

File: src/components/tooltip/tooltip.js (hunk 8 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: this is kind of constant the way it is now in the defaultProps. it is single place to update the defaults. i think it is clear the way it is now

File: src/components/tooltip/tooltip.js (hunk 9 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: is this still relevant?

File: src/components/tooltip/tooltip.js (hunk 10 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: fixed

File: src/components/tooltip/tooltip.js [hunk 13]
```diff
@@ -10,6 +36,123 @@ import {h, Component} from 'preact';
...
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): string {
+    let nextAlternativeType: string = null;
+    Object.keys(ToolTipType).forEach((key, index) => {
+      if (!nextAlternativeType && index > this.lastAlternativeTypeIndex && ToolTipType[key] !== this.state.type) {
+        this.lastAlternativeTypeIndex = index;
+        nextAlternativeType = ToolTipType[key];
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: done

File: src/components/tooltip/tooltip.js [hunk 14]
```diff
@@ -10,6 +37,117 @@ import {h, Component} from 'preact';
...
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): string {
+    const alternatives = Object.values(ToolTipType).filter(item => {
+      return item != this.props.type;
+    });
+    this.lastAlternativeTypeIndex = Math.min(this.lastAlternativeTypeIndex + 1, alternatives.length - 1);
+    return ((alternatives[this.lastAlternativeTypeIndex]: any): string);
```
Comment (yairans) [inline src/components/tooltip/tooltip.js]:
```
return Object.values(ToolTipType).find((item, index) => {
    if (index > this.lastAlternativeTypeIndex && item != this.props.type) {
        this.lastAlternativeTypeIndex = index;
        return true;
    } else {
        return false;
    }
});
```

File: src/components/shell/shell.js [hunk 15]
```diff
@@ -214,6 +214,9 @@ class Shell extends Component {
     const {isIPadOS, isTablet, isMobile} = player.env;
     this.props.updateIsMobile(isIPadOS || isTablet || isMobile || forceTouchUI);
     this._onWindowResize();
+    window.addEventListener('resize', () => {
```
Comment (yairans) [inline src/components/shell/shell.js]:
```suggestion
    this.props.eventManager.listen(window, player.Event.RESIZE, () => this._onWindowResize());
```

File: src/components/tooltip/tooltip.js [hunk 16]
```diff
@@ -10,6 +37,123 @@ import {h, Component} from 'preact';
...
+   * checks if after the render the tooltip is within boundaries of the player
+   * if not it will try to set a new type which will be checked after the next render
+   * @param {Object} prevProps - previous component props
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentDidUpdate(prevProps: Object): void {
+    if (this.props.playerClientRect !== prevProps.playerClientRect) {
+      this.lastAlternativeTypeIndex = -1;
+      this.setState({type: this.props.type});
```
Comment (yairans) [inline src/components/tooltip/tooltip.js]:
```suggestion
      this.setState({valid: false, type: this.props.type});
```
as you said yesterday. no?

File: src/components/shell/shell.js (hunk 15 above)
Comment (RoyBregman) [reply src/components/shell/shell.js]: fixed, thanks

File: src/components/tooltip/tooltip.js (hunk 16 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: yes, you actually listen to what i say :))) thanks!

File: src/components/shell/shell.js (hunk 15 above)
Comment (yairans) [inline src/components/shell/shell.js]: also see https://github.com/kaltura/playkit-js-ui/pull/367 and verify this issue with your change

Comment (OrenMe) [reply src/components/shell/shell.js]: use the event manager for this as well

File: src/components/shell/shell.js [hunk 17]
```diff
@@ -214,7 +214,10 @@ class Shell extends Component {
     const {isIPadOS, isTablet, isMobile} = player.env;
     this.props.updateIsMobile(isIPadOS || isTablet || isMobile || forceTouchUI);
     this._onWindowResize();
-    this.props.eventManager.listen(player, player.Event.RESIZE, () => this._onWindowResize());
+    this.props.eventManager.listen(window, 'resize', () => {
```
Comment (Yuvalke) [inline src/components/shell/shell.js]: why did you change it to window?

File: src/components/tooltip/tooltip.js [hunk 18]
```diff
@@ -10,6 +37,123 @@ import {h, Component} from 'preact';
...
+    this._hoverTimeout = null;
+  }
+
+  /**
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): ?string {
+    return ((Object.values(ToolTipType).find((item, index) => {
```
Comment (Yuvalke) [inline src/components/tooltip/tooltip.js]: `const ToolTipTypeValues = Object.values(ToolTipType)
ToolTipTypeValues.length > this.lastAlternativeTypeIndex ? this.lastAlternativeTypeIndex++ : this.lastAlternativeTypeIndex
return ToolTipTypeValues[this.lastAlternativeTypeIndex]`

File: src/components/tooltip/tooltip.js [hunk 19]
```diff
@@ -10,6 +37,123 @@ import {h, Component} from 'preact';
...
+    maxWidth: '240px'
+  };
+
+  /**
+   * @constructor
+   * @param {*} props props
+   * @memberof Tooltip
+   */
+  constructor(props: Object) {
+    super(props);
```
Comment (OrenMe) [inline src/components/tooltip/tooltip.js]: no need for constructor if there's no implementation in it

File: src/components/tooltip/_tooltip.scss (hunk 11 above)
Comment (RoyBregman) [reply src/components/tooltip/_tooltip.scss]: not possible as we discussed

File: src/components/tooltip/_tooltip.scss [hunk 20]
```diff
@@ -1,31 +1,129 @@
...
-    height: 10px;
-    background-color: #fff;
-    border-radius: 3px;
-    transform: rotate(45deg) translate3d(0%, 50%, 0);
-    left: 50%;
-    bottom: -1px;
+    line-height: normal;
+    font-weight: bold;
+    font-size: 13px;
+    z-index: 2;
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/components/tooltip/_tooltip.scss]: done

File: src/components/tooltip/tooltip.js [hunk 21]
```diff
@@ -10,6 +36,123 @@ import {h, Component} from 'preact';
...
+    this._hoverTimeout = null;
+  }
+
+  /**
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): string {
+    let nextAlternativeType: string = null;
```
[the remark this answers is not in the document]
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: outdated

File: src/components/tooltip/tooltip.js (hunk 19 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: removed

File: src/components/tooltip/tooltip.js (hunk 18 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: i am leaving it as it is

File: src/components/shell/shell.js (hunk 15 above)
Comment (RoyBregman) [reply src/components/shell/shell.js]: done

File: src/components/tooltip/tooltip.js (hunk 14 above)
Comment (RoyBregman) [reply src/components/tooltip/tooltip.js]: fixed
